pkg/operator/encryption/shared_encryption.go

```diff
@@ -305,21 +328,43 @@ func getAPIServerRevision(podClient corev1client.PodInterface) (string, error) {
 		case corev1.PodUnknown:
 			return "", fmt.Errorf("api server pod %s in unknown phase", apiServerPod.Name)
 		case corev1.PodSucceeded, corev1.PodFailed:
-			// old pods that we can safely ignore
+			// handle failed pods carefully to make sure things are healthy
+			// since the API server should never exit, a succeeded pod is considered as failed
+			failed.Insert(apiServerPod.Labels[revisionLabel])
 		default:
-			// do not error in case new phases get added
-			klog.Infof("api server pod %s has unexpected phase %v", apiServerPod.Name, phase)
+			// error in case new unexpected phases get added
+			return "", fmt.Errorf("api server pod %s has unexpected phase %v", apiServerPod.Name, phase)
 		}
 	}
 
 	if len(revisions) != 1 {
 		return "", nil // api servers have not converged onto a single revision
 	}
 	revision, _ := revisions.PopAny()
+
+	if failed.Has(revision) {
+		return "", fmt.Errorf("api server revision %s has both running and failed pods", revision)
+	}
+
+	revisionNum, err := strconv.Atoi(revision)
+	if err != nil {
+		return "", fmt.Errorf("api server has invalid revision: %v", err)
+	}
+
+	for _, failedRevision := range failed.List() { // iterate in defined order
+		failedRevisionNum, err := strconv.Atoi(failedRevision)
+		if err != nil {
+			return "", fmt.Errorf("api server has invalid failed revision: %v", err)
+		}
+		if failedRevisionNum > revisionNum { // TODO can this dead lock?
+			return "", fmt.Errorf("api server has failed revision %v which is newer than running revision %v", failedRevisionNum, revisionNum)
+		}
+	}
+
 	return revision, nil
```

Comment:
does it mean we want to tolerate failed pods with `failedRevisionNum` < `revisionNum`?